pkg/flowaggregator/s3uploader/s3uploader.go

```diff
@@ -0,0 +1,453 @@
+		}
+		p.region = region
+		p.awsS3Client = s3.NewFromConfig(cfg)
+		p.awsS3Uploader = s3manager.NewUploader(p.awsS3Client)
+	}
+	return nil
+}
+
+func (p *S3UploadProcess) SetUploadInterval(uploadInterval time.Duration) {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+	p.uploadInterval = uploadInterval
+	if p.uploadTicker != nil {
+		p.uploadTicker.Reset(p.uploadInterval)
+	}
+}
+
+func (p *S3UploadProcess) CacheRecord(record ipfixentities.Record) {
+	p.queueMutex.Lock()
+	defer p.queueMutex.Unlock()
+	r := flowrecord.GetFlowRecord(record)
+	p.writeRecordToBuffer(r)
+	// If the number of pending records in the buffer reaches maxRecordPerFile,
+	// add the buffer to bufferQueue.
+	if int32(p.cachedRecord) == p.maxRecordPerFile {
+		p.appendBufferToQueue()
+	}
+}
+
+func (p *S3UploadProcess) Start() {
+	p.startExportProcess()
+}
+
+func (p *S3UploadProcess) Stop() {
+	p.stopExportProcess(true)
+}
+
+func (p *S3UploadProcess) startExportProcess() {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+	if p.exportProcessRunning {
+		return
+	}
+	p.exportProcessRunning = true
+	p.uploadTicker = time.NewTicker(p.uploadInterval)
+	p.stopCh = make(chan stopPayload, 1)
+	p.exportWg.Add(1)
+	go func() {
+		defer p.exportWg.Done()
+		p.flowRecordPeriodicCommit()
+	}()
+}
+
+func (p *S3UploadProcess) stopExportProcess(flushQueue bool) {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+	if !p.exportProcessRunning {
+		return
+	}
+	p.exportProcessRunning = false
+	defer p.uploadTicker.Stop()
+	p.stopCh <- stopPayload{
+		flushQueue: flushQueue,
+	}
+	p.exportWg.Wait()
+}
+
+func (p *S3UploadProcess) flowRecordPeriodicCommit() {
+	klog.InfoS("Starting S3 exporting process")
+	ctx := context.Background()
+	for {
+		select {
+		case stop := <-p.stopCh:
+			klog.InfoS("Stopping S3 exporting process")
+			if !stop.flushQueue {
+				return
+			}
+			ctx, cancelFn := context.WithTimeout(ctx, bufferFlushTimeout)
+			defer cancelFn()
+			err := p.batchUploadAll(ctx)
+			if err != nil {
+				klog.ErrorS(err, "Error when doing batchUploadAll on stop")
+			}
+			return
+		case <-p.uploadTicker.C:
+			err := p.batchUploadAll(ctx)
+			if err != nil {
+				klog.ErrorS(err, "Error when doing batchUploadAll on triggered timer")
+			}
+		}
+	}
+}
+
+// batchUploadAll uploads all buffers cached in bufferQueue and previous fail-
+// to-upload buffers stored in buffersToUpload. Returns error if encountered.
+func (p *S3UploadProcess) batchUploadAll(ctx context.Context) error {
+	p.queueMutex.Lock()
+	if p.cachedRecord == 0 && len(p.bufferQueue) == 0 {
+		p.queueMutex.Unlock()
+		return nil
+	}
+	if p.cachedRecord != 0 {
+		p.appendBufferToQueue()
+	}
+	// dump cached buffers from bufferQueue to buffersToUpload
+	for _, buf := range p.bufferQueue {
+		p.buffersToUpload = append(p.buffersToUpload, buf)
+		if len(p.buffersToUpload) > maxKeptBufferNum {
+			p.buffersToUpload = p.buffersToUpload[1:]
+		}
+	}
+	p.bufferQueue = p.bufferQueue[:0]
+	p.queueMutex.Unlock()
```

Comment:
please use this style as it makes lock management cleaner (it also takes care of my previous comment about returning)

```go
func() {
        p.queueMutex.Lock()
        defer p.queueMutex.Unlock()
        // ...
}()
```